File: greedy/min_refueling_stops.py
import heapq


def compute_min_stops(target: int, start_fuel: int, stations: list[list[int]]) -> int:  # LeetCode Q.871.
    stops, fuel_max_heap = 0, []
    current_position, current_fuel = start_fuel, 0
    while current_position < target:
        while stations and current_position >= stations[0][0]:  # Visitable stations.
            heapq.heappush(fuel_max_heap, -stations.pop(0)[1])  # Negate for max heap.

        while current_fuel <= 0:
            if not fuel_max_heap:  # Past gas stations are all visited.
                return -1
            current_fuel -= heapq.heappop(fuel_max_heap)
            stops += 1

        current_position += current_fuel
        current_fuel -= current_fuel  # All fuel is used to forward position.

    return stops

File: greedy/min_refueling_stops/min_refueling_stops.cpp
#include <vector>
#include <algorithm>
using namespace std;

int minRefuelStops(int target, int startFuel, vector<vector<int>> &stations)
{ // LeetCode Q.871.
    int minStops = 0;
    int currentPosition = startFuel;

    startFuel -= startFuel;
    vector<int> fuelMaxHeap;

    while (currentPosition < target)
    {
        // Collect all the visited gas stations.
        while (!stations.empty() && currentPosition >= stations[0][0])
        {
            fuelMaxHeap.push_back(stations[0][1]);
            push_heap(fuelMaxHeap.begin(), fuelMaxHeap.end());
            stations.erase(stations.begin());
        }

        if (startFuel == 0) // Time to refuel.
        {
            if (fuelMaxHeap.empty())
                return -1;

            startFuel += fuelMaxHeap[0];
            minStops += 1;
            pop_heap(fuelMaxHeap.begin(), fuelMaxHeap.end());
            fuelMaxHeap.pop_back();
        }

        currentPosition += startFuel; // Spend all the fuel to move on.
        startFuel -= startFuel;
    }

    return minStops;
}
